src/darwin/list.mm

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <nan.h>
+#include "../drivelist.hpp"
+
+#import <Cocoa/Cocoa.h>
+#import <DiskArbitration/DiskArbitration.h>
+
+namespace Drivelist {
+
+  NSNumber *DictNum(CFDictionaryRef dict, const void *key) {
+      return (NSNumber*)CFDictionaryGetValue(dict, key);
+  }
+
+  std::vector<DeviceDescriptor> ListStorageDevices() {
+    std::vector<DeviceDescriptor> deviceList;
+
+    NSArray* volumeKeys = [NSArray arrayWithObjects:NSURLVolumeNameKey, NSURLVolumeIsRemovableKey, nil];
+    NSArray* volumePaths = [[NSFileManager defaultManager]
+                            mountedVolumeURLsIncludingResourceValuesForKeys:volumeKeys
```

Comment:
@lurch I think that you might be right, I will do some testing with unmounted storage devices and get back with the results.

Reply:
Do not have access to a storage device that I can wipe yet, but I experimented a bit with using DiskArbitration to list the devices from a disk perspective instead of using the volume approach.

The good thing is that it became a bit more natural since the DeviceDescriptor will be built on a device level and does not need to be merged, instead mount points will be added to the devices afterwards.

The ugly part is that it involved some black magic to force DiskArbitration to run sync, but I can abstract that away.